[WIP] tests: Migrate legacy test 32 not working

nelsyeung opened the pull request:
I'm attempting to migrate test 32 as per issue #1286, but I'm hitting a small road block.
I've tried writing a simplified version just to test smaller parts, but I realised that it's failing already with just the first 5 lines. Also tried just doing the commands in nvim, but no luck. Even vim doesn't seem to follow the expects.

Here's the output from my test:
![screen shot 2015-06-25 at 23 40 32](https://cloud.githubusercontent.com/assets/3951541/8367245/42bfb0ca-1b94-11e5-950c-fc6d680b4a3d.png)

Any advice would be greatly appreciated.
Thank you very much.

Commit:
tests: Migrate legacy test 32

Commit:
tests: Migrate legacy test 32

justinmk commented on test/functional/legacy/032_insert_expansion_spec.lua:
```diff
@@ -0,0 +1,90 @@
+    execute('se cpt=.,w ff=unix | $-2,$w!Xtestfile | set ff&')
+    execute('se cot=')
+    feed('<C-W>nO#include "Xtestfile"<cr>')
+    feed('ru<C-N><C-N><C-X><C-N><esc><C-A><cr>')
```
Possibly the `<esc><c-a>` is being interpreted as "meta ctrl-a" chord. Need to break it up.

nelsyeung replied:
> Possibly the <esc><c-a> is being interpreted as "meta ctrl-a" chord. Need to break it up. Same for the other <esc> feedkeys below.

I don't think that's the problem, I've tried breaking them up but no luck. In addition, since I've tried to run those commands by hand in both Vim and nvim, there's no way it's a mistake with the code but the commands itself. Thus, I feel that the original test is incorrect, though this doesn't make any sense as well.

Some examples in which I think the original test is flawed:

1. Logically if we run `ru<c-n><c-n>` we should get `run2`. However, if you look at the expect or test32.ok you will see `run1 run3`. Removing the second `<c-n>` (i.e., `ru<c-n>`), however, will work as expected.
2. For `execute('se cpt=u nohid')
feed('<c-w>oOEN<c-n><cr>')
feed('unl<c-n><c-x><c-x><c-p><esc><cr>')`,
following the logic, none of the unloaded buffers will contain the text `ENDTEST` or `unless`, because during the previous write command it's on the wrong window, therefore the autocomplete will not work for these two lines. I can get around this by adding `feed('<c-w>j<cr>')` before the writes.

I have yet to find the flaw on the last 3 lines of the test. If I can find it then this test should be done. But am I allowed to add more and remove commands to the original test like this? And can the original test really be wrong?

justinmk replied:
@lucc may have some insight... I would hesitate to change the test expectations without being 100% sure I know why I'm doing it.

justinmk commented on test/functional/legacy/032_insert_expansion_spec.lua:
```diff
@@ -0,0 +1,90 @@
+
+    execute('se backspace=""')
+    execute('se cpt=.,w ff=unix | $-2,$w!Xtestfile | set ff&')
+    execute('se cot=')
```
In general, prefer to use the long option name instead of the abbreviated name.

lucc commented:
> Logically if we run `ru<c-n><c-n>` we should get `run2`. However, if you look at the expect or test32.ok you will see `run1 run3`. Removing the second `<c-n>` (i.e., `ru<c-n>`), however, will work as expected.

The `<C-A>` in line 44 should turn `run2` into `run3` I think.

I think another problem with this test is that the original test has the test32.in file loaded into a buffer. If I add
```
:redir > foo
:ls!
:redir END
```
at the end of `test32.in` (before the `:wq`) and run `make -C src/nvim/testdir TESTNUM=32` I get this output in `foo`:
```
  1      "test32.in"                    line 58
  2u     "Xtestfile"                    line 1
  3 %a + "Xtest11.one"                  line 14
  5u#    "Xtest11.two"                  line 1
```
This might be relevant if we have `set cpt=.,w` or `cpt=u`.

lucc commented:
Is there any kind of nvim command which allows us to dump a large part of the editor state to a file in order to simplify things like above? I am thinking about a simple `:dumpdebuginfo domefile`. Then we can add that to the original test file to find out what is going on. The command should dump buffer, window and tab information and such, and registers and mode and so on.

nelsyeung commented:
> The <C-A> in line 44 should turn run2 into run3 I think.

No, the problem is that line 44 `ru<c-n><c-n>` will create: `run2`, so the next command `<c-x><c-n>`will give `end`, thus `run2 end` ≠ `run1 run3`. Again, if I change line 44 to `ru<c-n><c-x><c-n><esc><c-a>` then it will work correctly, but that's changing the original test.

> I think another problem with this test is that the original test has the test32.in file loaded into a buffer. If I add

I did realise that, I have tried just putting everything in test32.in into `insert`, but no luck. I will try your suggestion with `make -C src/nvim/testdir TESTNUM=32` myself and see what's happening.

Thanks

lucc commented:
Are you still working on this? I started working on it myself and got it nearly fixed. Only a small part is missing. Would you like to include my work or should I just open a new PR to discuss this?

nelsyeung commented:
Oh I haven't been working on this but I'm rather interested at how you've fixed it. I guess including it would be nicer since it's one less pull request?

lucc commented:
see https://github.com/lucc/neovim/tree/test32

nelsyeung commented:
"Only a small part is missing" - you've added quite a bit to me actually, haha. I've squashed and pushed your changes now. Please check.

mhinz committed:
Syntax: skip over ]

The ported patch for 7.4.825 removed code that is needed to walk over "]" by
accident.

0: https://github.com/neovim/neovim/commit/52692d3cd3e682a4116d3cec1fcf05880f0c77a1
1: https://github.com/vim/vim/commit/1560d07045d416d0abf9731c43c28925f61515b6

Fixes #4207.

lucc commented:
Would you mind rebasing on current master? Makes it easier to see the differences between our versions.

Also I think it is better to not squash things until the review is done. Afterwards I tend to sqash into *few meaningful* and understandable commits, not *one*.

Commit:
Merge branch 'test32' of https://github.com/nelsyeung/neovim into test32

justinmk commented:
test32 was removed in 2de447b60c04fcaad3d762851c512b2545ea6bad
